engine/src/test/java/io/camunda/zeebe/engine/state/distribution/DistributionStateTest.java

```diff
@@ -0,0 +1,167 @@
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.distribution;
+
+import static io.camunda.zeebe.util.buffer.BufferUtil.wrapString;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.camunda.zeebe.engine.state.mutable.MutableDistributionState;
+import io.camunda.zeebe.engine.state.mutable.MutableProcessingState;
+import io.camunda.zeebe.engine.util.ProcessingStateExtension;
+import io.camunda.zeebe.model.bpmn.Bpmn;
+import io.camunda.zeebe.protocol.impl.record.value.deployment.DeploymentRecord;
+import io.camunda.zeebe.protocol.impl.record.value.distribution.CommandDistributionRecord;
+import io.camunda.zeebe.protocol.record.ValueType;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ProcessingStateExtension.class)
+public final class DistributionStateTest {
+  private MutableProcessingState processingState;
+  private MutableDistributionState distributionState;
+
+  @BeforeEach
+  public void setUp() {
+    distributionState = processingState.getDistributionState();
+  }
+
+  @Test
+  public void shouldReturnFalseOnEmptyStateForHasPendingCheck() {
+    // when
+    final var hasPending = distributionState.hasPendingDistribution(10L);
+
+    // then
+    assertThat(hasPending).isFalse();
+  }
+
+  @Test
+  public void shouldAddPendingDistribution() {
+    // given
+    final var distributionKey = 10L;
+    final var partition = 1;
+
+    // when
+    distributionState.addPendingDistribution(distributionKey, partition);
+
+    // then
+    assertThat(distributionState.hasPendingDistribution(distributionKey)).isTrue();
+  }
+
+  @Test
+  public void shouldRemovePendingDistribution() {
+    // given
+    final var deploymentKey = 10L;
+    final var partition = 1;
+    distributionState.addPendingDistribution(deploymentKey, partition);
+
+    // when
+    distributionState.removePendingDistribution(deploymentKey, partition);
+
+    // then
+    assertThat(distributionState.hasPendingDistribution(deploymentKey)).isFalse();
+  }
+
+  @Test
+  public void shouldReturnNullOnRequestingStoredDistributionWhenNothingStored() {
+    // when
+    final var distributionRecord = distributionState.getCommandDistributionRecord(1, 1);
+
+    // then
+    assertThat(distributionRecord).isNull();
+  }
+
+  @Test
+  public void shouldStoreDistributionInState() {
+    // given
+    final var distributionRecord = createCommandDistributionRecord();
+
+    // when
+    distributionState.addCommandDistribution(1, distributionRecord);
+
+    // then
+    final var storedDistribution =
+        distributionState.getCommandDistributionRecord(1, distributionRecord.getPartitionId());
```

Comment:
❓ Why can we retrieve a command distribution record based on the key **and** the partition id? The command distribution record doesn't change between the partitions, right? So, I thought we only have to store it once (during `STARTED`) and can remove it once all pending distributions are `ACKNOWLEDGED` (remove it when `FINISHED`).

Reply:
From my understanding if we use a composite key here this means we would have to store the command distribution for each partition, instead of only once. This isn't necessary because as you mention, each partition uses the same record anyways.

Now it is:
```
distribution key => value type & record
```

Otherwise it would become
```
distribution key & partition id 1 => value type & record
distribution key & partition id 2 => value type & record
distribution key & partition id 3 => value type & record
...
```

Reply:
Sorry I think I misunderstood your question here. I ask for the id because of [this](https://github.com/camunda/zeebe/pull/11869#discussion_r1121711934)